File: app/src/main/java/com/example/khughes/machewidget/VehicleViewModel.java
package com.example.khughes.machewidget;

import android.app.Application;

import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.example.khughes.machewidget.db.VehicleInfoDao;
import com.example.khughes.machewidget.db.VehicleInfoDatabase;

import java.util.ArrayList;
import java.util.List;

public class VehicleViewModel extends AndroidViewModel {

    private final VehicleRepository mRepository;
    private final LiveData<List<VehicleIds>> mAllVehicles;

    public VehicleViewModel (Application application) {
        super(application);
        mRepository = new VehicleRepository(application);
        mAllVehicles = mRepository.getAllVehicles();
    }

    LiveData<List<VehicleIds>> getAllVehicles() { return mAllVehicles; }

    public void setEnable(String VIN, boolean value) { mRepository.enable(VIN,value);}

    public int countEnabledVehicle() {
        int count = 0;
        for(VehicleIds id: mAllVehicles.getValue()) {
            if(id.isEnabled()) {
                ++count;
            }
        }
        return count;
    }

    class VehicleRepository {

        private final VehicleInfoDao mVehDao;
        private final LiveData<List<VehicleIds>> mAllVehicles;

        VehicleRepository(Application application) {
            mVehDao = VehicleInfoDatabase.getInstance(application).vehicleInfoDao();
            mAllVehicles = mVehDao.getLiveDataVehicleInfo();
//            ArrayList<VehicleIds> x = new ArrayList<>();
//            VehicleIds one = new VehicleIds();
//            one.setVIN("3FMTK3R75MMA09929");
//            one.setEnabled(true);
//            one.setNickname("Howard");
//            x.add(one);
//            VehicleIds two = new VehicleIds();
//            two.setVIN("3FMTK3R75MMA09930");
//            two.setEnabled(true);
//            two.setNickname("Raj");
//            x.add(two);
//            VehicleIds tre = new VehicleIds();
//            tre.setVIN("3000K3R75MMA09930");
//            tre.setEnabled(false);
//            tre.setNickname("Penny");
//            x.add(tre);
//            mAllVehicles = new MutableLiveData<>(x);
        }

        LiveData<List<VehicleIds>> getAllVehicles() {
            return mAllVehicles;
        }

        void enable(String VIN, boolean value) {
            new Thread( () -> mVehDao.updateEnable(VIN, value)).start();
        }
    }
}

File: app/src/main/java/com/example/khughes/machewidget/VehicleViewModel.kt
package com.example.khughes.machewidget

import android.app.Application
import android.content.Context
import androidx.lifecycle.AndroidViewModel
import androidx.lifecycle.viewModelScope
import com.example.khughes.machewidget.db.VehicleInfoDao
import com.example.khughes.machewidget.db.VehicleInfoDatabase
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.launch

class VehicleViewModel(application: Application?) : AndroidViewModel(application!!) {
    private val mRepository: VehicleRepository = VehicleRepository(application?.applicationContext as Context)
    val allVehicles: Flow<List<VehicleIds>>

    init {
        allVehicles = mRepository.allVehicles
    }

    fun setEnable(vehicleId: String, value: Boolean) {
        mRepository.enable(vehicleId, value)
    }

    fun setModel(vehicleId: String, value: Vehicle.Companion.Model) {
        viewModelScope.launch {
            mRepository.setModel(vehicleId, value)
        }
    }

    fun removeVehicle(vehicleId: String) {
        mRepository.remove(vehicleId)
    }

    internal inner class VehicleRepository(context: Context) {
        private val mVehDao: VehicleInfoDao = VehicleInfoDatabase.getInstance(context).vehicleInfoDao()
        private var mContext: Context
        var allVehicles: Flow<List<VehicleIds>>

        init {
            allVehicles = mVehDao.liveDataVehicleInfo
            mContext = context
        }

        fun enable(vehicleId: String, value: Boolean) {
            viewModelScope.launch {
                mVehDao.updateEnable(vehicleId, value)
            }
        }

        fun setModel(vehicleId: String, value: Vehicle.Companion.Model) {
            viewModelScope.launch(Dispatchers.IO) {
                mVehDao.updateModel(vehicleId, value)
//                mVehDao.liveDataVehicleInfo.collectLatest {
//                    allVehicles = it
//                }
            }
        }

        fun remove(vehicleId: String) {
            viewModelScope.launch{
                mVehDao.deleteVehicleInfoByVehicleId(vehicleId)
                allVehicles = mVehDao.liveDataVehicleInfo
                // Force recalculation of electric vehicles
                val info = InfoRepository(mContext)
                info.vehicles.size
            }
        }
    }
}